### PR TITLE
outputDataSchema addition

### PR DESCRIPTION
**Pull request by fjtirado**
This adds possibility of validating workflow output in a similar way that is already done for workflow input

See issue https://github.com/serverlessworkflow/specification/issues/611

Signed-off-by: Francisco Javier Tirado Sarti <ftirados@redhat.com>

**Many thanks for submitting your Pull Request :heart:!**

**Please specify parts of this PR update:**

- [x] Specification
- [x] Schema
- [ ] Examples
- [ ] Extensions
- [ ] Roadmap
- [ ] Use Cases
- [ ] Community
- [ ] TCK
- [ ] Other

**Discussion or Issue link**:

**What this PR does / why we need it**:

**Special notes for reviewers**:

**Additional information:**

### CODE REVIEW

**Comment by tsurdilo** [comment]
There is also question on what happens on failure to validate subflow result as this can be defined in the subflow workflow definition, so this is tied to the error handling still WIP pr. Imo they should be committed at the same time if we decided to go w/ this.

**Comment by tsurdilo** [comment]
related to https://github.com/serverlessworkflow/specification/pull/673, please do not merge one without the other

**Comment by fjtirado** [comment]
@tsurdilo I opened this PR because reading  @cdavernas  I believe this feature was accepted. 
In my view, the output validation should be performed only for parent process (not subprocess) and if it fails, it behave as when input validation fails: it throws an exception if the failOnValidation is true. 
Since it only applies to parent process, we can consider it is is not part of error handling feature, because it does not really belong to flow execution. Thats why I just change the dataInputShema bits to add dataOuputSchema.

**Comment by tsurdilo** [comment]
Yeah definitely lets add it, just would be nice to update the little things mentioned. Ty!!

**Comment by fjtirado** [comment]
> 

@tsurdilo Sure, Ill do, thanks

**Comment by fjtirado** [comment]
@tsurdilo I tried to reflect your concerns into the text.

File: specification.md
Comment (cdavernas) [inline specification.md:1868]: I would instead write something like:

Successfully means that the workflow has completed an end state without errors

**Review by cdavernas — CHANGES_REQUESTED** [review]
Please consider #705 before merging, as requested change might be included in this PR as well.

**Comment by fjtirado** [comment]
@cdavernas Comments incorporated

**Review by cdavernas — APPROVED** [review]
Looks great! Many thanks!

**Review by cdavernas — APPROVED** [review]

**Review by ricardozanini — APPROVED** [review]

File: schema/workflow.json
Comment (tsurdilo) [inline schema/workflow.json:107]: align with https://github.com/serverlessworkflow/specification/blob/main/schema/workflow.json#L69

File: schema/workflow.json
Comment (tsurdilo) [inline schema/workflow.json:79]: This seems a copy of "dataInputSchema"/ "inputDataSchema" as you changed.
Would be better imo to have a single definition called "inputSchema" and then reference it rather than having two separate

File: specification.md
Comment (tsurdilo) [inline specification.md:1870]: dont use "might", be specific. yes it should be propagated to parent exec and handled as part of the workflow state that invoked the subflow.

File: specification.md
Comment (tsurdilo) [inline specification.md:1911]: think it should be URI, not URL

File: specification.md
Comment (tsurdilo) [inline specification.md:1870]: "thrown" is kinda java-ish. pls use "raised"

**Review by tsurdilo — CHANGES_REQUESTED** [review]
Overall lgtm. Left some comments that should be fixed pls

File: schema/workflow.json
Comment (cdavernas) [reply schema/workflow.json:79]: @tsurdilo I would recommend calling that definition `dataValidation` or whatever, as `inputSchema` would be confusing when used for output 😜

File: specification.md
Comment (cdavernas) [reply specification.md:1911]: Agreed!

File: specification.md
Comment (cdavernas) [reply specification.md:1870]: Totally disagree here! Errors are not raised, they are thrown, it's not java-ish, it's IT-ish and is the proper terminology in that case imo

Comment (cdavernas) [reply specification.md:1870]: As we are speaking of error control flow, the terminology `catch` is actually more appropriate. 
As a matter of fact, raising an error is the right terminology when "producing" the error.

Btw, `throw`, `catch` and other keywords of the like are NOT specific to Java or a specific grammar. Those are specific IT terms, very widely and globally used, across languages and techs, to describe, in this case, error control flow.  It is not a matter of opinion, it's a matter of proper vocabulary. Making proper use of that vocabulary ensures ubiquitousness. Using synonyms for the sake of not looking like code is  IMHO very dangerous, misleading to the biggest part of our community (devs mostly) and is even confusing.

Comment (tsurdilo) [reply specification.md:1870]: there is no throw and catch in Go :)

File: schema/workflow.json
Comment (tsurdilo) [reply schema/workflow.json:79]: maybe inputOutputValidation then cause we are not validating like event payloads or action results or state data which are all data

File: specification.md
Comment (cdavernas) [reply specification.md:1870]: @tsurdilo indeed, because there is no proper error handling out of the box. 
 https://link.medium.com/nkJbfoMHuvb

Doesn't mean that because they do not provide a core feature that said feature is wrongly named 😜

File: schema/workflow.json
Comment (cdavernas) [reply schema/workflow.json:79]: @tsurdilo Point taken! However, I personally dislike that term too. I'm sure we can, however, find a good one. I'll sleep on it.

File: specification.md
Comment (tsurdilo) [reply specification.md:1870]: | They are basic IT terms

yes exactly, thats why we should if possible stay away from using them inside spec text. DSLs are to be read and understood by non-IT people too

Comment (cdavernas) [reply specification.md:1870]: But dude, catch and throw are less misleading that synonyms that do not apply to the concepts.
Again, it's not because something is rightfully used by global consensus in IT that we should not use it and confuse most ppl. 
Those terms apply perfectly, are easily understood by even the dumbest guy around, and avoids guaranteed misinterpretation.

Comment (tsurdilo) [reply specification.md:1870]: It would be also easier to use "return error" and "check error" and "panic" and "unwrap" too but i guess this maybe too much. So idk you seem to have a strong opinion about this so not going to argue with you, but don't think try/catch is really that "mainstream" outside devel world maybe

Comment (cdavernas) [reply specification.md:1870]: Well, I do have a very strong opinion on that matter, but that's just me, and I'm indeed a bit of a grammar nazy 😜 
Let's see what the others think about it, and let's settle this out in a weekly. 
Wdyt?

Comment (cdavernas) [reply specification.md:1870]: Suggested terms are all far fetched imho, and do not really paint the 'action' we want to describe, again imho.

Comment (tsurdilo) [reply specification.md:1870]: Sure, lets try that and catch any issues that come up ;)

Comment (cdavernas) [reply specification.md:1870]: Jokes appart, I hear you and I do believe you make a very valid point by always thinking of not turning the spec into some kind of IT only language. 
I just think that in this specific case, as the grammar nazy I must admit I am, alternatives sounds a bit far fetched whereas the try/catch/throw one is beautifully simple and ubiquitous. 
Let's see what the others think and settle that out in a weekly, wdyt?

Comment (fjtirado) [reply specification.md:1870]: I can use "generated" rather than "thrown", is that ok for you guys?

File: schema/workflow.json
Comment (fjtirado) [reply schema/workflow.json:79]: I think we can use "validationSchema" for the data stucture used to validate schema

File: specification.md
Comment (cdavernas) [reply specification.md:1870]: I don't think that's right either. I like the thrown term. But whathever, it's just a minor semantic detail, so I'll go with whatever you guys settle on 😉

File: schema/workflow.json
Comment (cdavernas) [reply schema/workflow.json:79]: I like that, good idea!

File: specification.md
Comment (cdavernas) [reply specification.md:1870]: I agree with @fjtirado on this one!

Comment (fjtirado) [reply specification.md:1870]: I do not have particular interest on discussing a semantic thing, so Tiho let us know which is your chosen word. 
If it helps, thow was not invented by Java, it was chosen for C++. And Java also includes (inherited from other languages) english words like if, else, finally, int, float, double, string, array... ;)

Comment (ricardozanini) [reply specification.md:1870]: I'd rather stay with the ubiquitous term, which are try/catch/throw, widely used by the IT community. Raised is yet another good term, thou. But in this context, throw is the best one. And just to try to add a period in this conversation: https://english.stackexchange.com/questions/319126/why-exactly-is-the-verb-to-throw-used-when-an-error-occurs-in-the-program.

**Comment by fjtirado** [comment]
@tsurdilo Comments incorporated, please take a look

**Comment by cdavernas** [comment]
@tsurdilo Can you please review requested changes?

**Comment by cdavernas** [comment]
Linked to https://github.com/serverlessworkflow/specification/issues/705

**Comment by ricardozanini** [comment]
@tsurdilo can you please take a final look so we can merge it? Thank you!

File: specification.md
Comment (tsurdilo) [inline specification.md:1868]: I think this sentence is confusing as in "completed an end state without errors" I think this should just say

"It's performed on the output of the workflow execution"..leave it at that imho

File: specification.md
Comment (tsurdilo) [inline specification.md:1870]: please change "might" to "can be" or "should be"

File: specification.md
Comment (tsurdilo) [inline specification.md:1872]: "setting failiOnvalidationError to false to not throw error even in case of validation errors"

i think would sound better

**Comment by tsurdilo** [comment]
left couple of small change requests for the docs part hope thats ok

File: specification.md
Comment (fjtirado) [reply specification.md:1870]: Done

File: specification.md
Comment (fjtirado) [reply specification.md:1872]: Done

**Comment by tsurdilo** [comment]
I see, i think both says they are the same, they are both schemas one defines data input and the other data output of the workflow exec

**Comment by ricardozanini** [comment]
@tsurdilo @cdavernas @fjtirado I'd rather stay with `dataInputSchema` and `dataOutputschema` if you don't mind. @tsurdilo I think all your comments have been handled. Can you take a final look/approve so we can merge this?

**Comment by cdavernas** [comment]
> @tsurdilo @cdavernas @fjtirado I'd rather stay with `inputDataSchema` and `outputDataSchema` if you don't mind.

Works for me either way! I guess it's just a matter of point of view/personal prefs anyways

**Comment by ricardozanini** [comment]
@cdavernas I agree. Either works for me too if we can move forward!

**Comment by fjtirado** [comment]
@tsurdilo Any further request?

File: roadmap/README.md
Comment (tsurdilo) [inline roadmap/README.md:48]: so which one is it? dataOutputSchema as requested or inputDataSchema as in the change you made to workflow.json :)

Comment (fjtirado) [reply roadmap/README.md:48]: If I revert to dataInputSchema, I guess thats fine isnt it?

Comment (tsurdilo) [reply roadmap/README.md:48]: imho yes, but depends if you and others want to do that change. either way after you guys decide 
please make sure that its reflected in readme.md with the same names you use in workflow schema

Comment (fjtirado) [reply roadmap/README.md:48]: @tsurdilo   I revert to the original form, dataInputSchema and dataOutputSchema so it is aligned with readme roapmap.

**Review by tsurdilo — APPROVED** [review]

**Comment by ricardozanini** [comment]
@cdavernas any other comments?

**Comment by cdavernas** [comment]
@ricardozanin Nope, looking great to me! Thanks @fjtirado

**Pull request merged**
